src/v/datalake/protobuf_to_arrow_converter.cc

```diff
@@ -0,0 +1,126 @@
+#include "datalake/protobuf_to_arrow_converter.h"
+
+#include "datalake/proto_to_arrow_struct.h"
+
+#include <arrow/type.h>
+
+#include <memory>
+#include <stdexcept>
+
+datalake::proto_to_arrow_converter::proto_to_arrow_converter(
+  std::string schema) {
+    initialize_protobuf_schema(schema);
+    if (!initialize_struct_converter()) {
+        // FIXME: exception type
+        throw std::runtime_error("Could not initialize arrow arrays");
+    }
+}
+void datalake::proto_to_arrow_converter::add_message(
+  const std::string& serialized_message) {
+    std::unique_ptr<google::protobuf::Message> message = parse_message(
+      serialized_message);
+    if (message == nullptr) {
+        // TODO: Return error so higher-level code can handle failed
+        // messages
+        return;
+    }
+
+    if (!_struct_converter->add_top_level_message(message.get()).ok()) {
+        // TODO: Error handling
+        throw new std::runtime_error("Could not add value");
+    }
+}
+void datalake::proto_to_arrow_converter::finish_batch() {
+    if (!_struct_converter->finish_batch().ok()) {
+        throw new std::runtime_error("Could not finish batch");
+    }
+}
+std::shared_ptr<arrow::Table>
+datalake::proto_to_arrow_converter::build_table() {
+    auto table_result = arrow::Table::FromChunkedStructArray(
+      _struct_converter->finish());
+    if (table_result.ok()) {
+        return table_result.ValueUnsafe();
+    } else {
+        return nullptr;
+    }
+}
+std::vector<std::shared_ptr<arrow::Field>>
+datalake::proto_to_arrow_converter::build_field_vec() {
+    return _struct_converter->get_field_vector();
+}
+std::shared_ptr<arrow::Schema>
+datalake::proto_to_arrow_converter::build_schema() {
+    return arrow::schema(_struct_converter->get_field_vector());
+}
+
+void datalake::proto_to_arrow_converter::initialize_protobuf_schema(
+  const std::string& schema) {
+    google::protobuf::io::ArrayInputStream proto_input_stream(
+      schema.c_str(), schema.size());
+    google::protobuf::io::Tokenizer tokenizer(&proto_input_stream, nullptr);
+
+    google::protobuf::compiler::Parser parser;
+    if (!parser.Parse(&tokenizer, &_file_descriptor_proto)) {
+        // TODO: custom exception type, or does something exist in wasm or
+        // schema registry already?
+        throw std::runtime_error("Could not parse protobuf schema");
+    }
+
+    if (!_file_descriptor_proto.has_name()) {
+        _file_descriptor_proto.set_name("dummy_message_name");
+    }
+
+    _file_desc = _protobuf_descriptor_pool.BuildFile(_file_descriptor_proto);
```

Comment:
let's hook up a protobuf::ErrorCollector here and use the other build file interface so we can redirect errors to our logs.